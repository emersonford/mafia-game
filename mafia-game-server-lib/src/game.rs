//! Core logic for a game of Mafia.

use std::collections::HashMap;
use std::collections::HashSet;
use std::time::Duration;
use std::time::SystemTime;

use mafia_game_lib::Allegiance;
use mafia_game_lib::ClientId;
use mafia_game_lib::Cycle;
use mafia_game_lib::PlayerStatus;
use mafia_game_lib::SpecialRole;
use rand::Rng;
use rand::seq::SliceRandom;
use tracing::field;

use crate::client::ClientState;
use crate::error::MafiaGameError;

// TODO(emersonford): allow this to be populated at runtime
#[derive(Clone, Debug)]
pub struct GameConfig {
    pub start_cycle: Cycle,
    pub time_for_day: Duration,
    /// End the day cycle early if all votes have been submitted.
    pub end_day_after_all_votes: bool,
    pub time_for_night: Duration,
    /// End the night cycle early if all votes have been submitted.
    pub end_night_after_all_votes: bool,
    pub num_special_roles: HashMap<SpecialRole, usize>,
    /// Time after cycle start during switch votes are rejected.
    ///
    /// Useful to avoid last-minute votes leaking into the next cycle and spoiling results.
    pub vote_grace_period: Duration,
    // TODO(emersonford): add option to reveal roles on death
}

/// State for an active game.
pub(crate) struct Game {
    config: GameConfig,
    role_to_players: HashMap<SpecialRole, Vec<ClientId>>,
    player_to_role: HashMap<ClientId, SpecialRole>,
    player_status: HashMap<ClientId, PlayerStatus>,
    cycle: Cycle,
    day_num: usize,
    cycle_start: SystemTime,
    /// Map of voter -> who they are voting for.
    ///
    /// If value is `None`, means the voter skipped voting.
    votes: HashMap<ClientId, Option<ClientId>>,
}

impl Game {
    pub(crate) fn start<S: Rng>(
        config: GameConfig,
        clients: &ClientState,
        mut seed: S,
    ) -> Result<Self, MafiaGameError> {
        let mut clients = clients.list_clients().values().copied().collect::<Vec<_>>();
        // Sort for determinism with deterministic seed.
        clients.sort();

        let num_mafia_roles = config
            .num_special_roles
            .get(&SpecialRole::Mafia)
            .copied()
            .unwrap_or(0);
        let total_special_roles = config.num_special_roles.values().copied().sum::<usize>();

        if num_mafia_roles == 0 {
            return Err(MafiaGameError::InvalidGameConfig(
                "need at least 1 mafia, got 0".to_string(),
            ));
        }

        if num_mafia_roles * 2 >= clients.len() {
            return Err(MafiaGameError::NotEnoughPlayers(format!(
                "need at least {} players to play with {} mafia, only have {} players",
                num_mafia_roles * 2 + 1,
                num_mafia_roles,
                clients.len()
            )));
        }

        if total_special_roles > clients.len() {
            return Err(MafiaGameError::NotEnoughPlayers(format!(
                "{} special roles were provided, but only have {} players",
                total_special_roles,
                clients.len()
            )));
        }

        let mut num_special_roles = config
            .num_special_roles
            .clone()
            .into_iter()
            .collect::<Vec<_>>();
        // Sort for determinism with deterministic seed.
        num_special_roles.sort_by_key(|(role, _)| *role);

        clients.shuffle(&mut seed);

        let mut role_to_players = HashMap::new();
        let mut player_to_role = HashMap::new();

        let mut client_idx = 0;
        for (special_role, num) in num_special_roles {
            for _ in 0..num {
                let client_id = clients[client_idx];

                role_to_players
                    .entry(special_role)
                    .or_insert_with(Vec::new)
                    .push(client_id);
                player_to_role.insert(client_id, special_role);

                client_idx += 1;
            }
        }

        let cycle = config.start_cycle;

        Ok(Game {
            config,
            role_to_players,
            player_to_role,
            player_status: clients
                .into_iter()
                .map(|client_id| (client_id, PlayerStatus::Alive))
                .collect(),
            cycle,
            day_num: 0,
            cycle_start: SystemTime::now(),
            votes: HashMap::new(),
        })
    }

    #[cfg(test)]
    pub(crate) fn get_player_roles(&self) -> &HashMap<ClientId, SpecialRole> {
        &self.player_to_role
    }

    pub(crate) fn get_player_status(&self, client_id: ClientId) -> PlayerStatus {
        self.player_status
            .get(&client_id)
            .copied()
            .unwrap_or(PlayerStatus::NotPlaying)
    }

    pub(crate) fn get_player_role(&self, client_id: ClientId) -> Option<SpecialRole> {
        self.player_to_role.get(&client_id).copied()
    }

    pub(crate) fn get_player_allegiance(&self, client_id: ClientId) -> Allegiance {
        self.get_player_role(client_id)
            .map_or(Allegiance::Villagers, |role| role.allegiance())
    }

    pub(crate) fn get_alive_players(&self) -> impl Iterator<Item = ClientId> {
        self.player_status.iter().filter_map(|(client_id, status)| {
            if *status == PlayerStatus::Alive {
                Some(*client_id)
            } else {
                None
            }
        })
    }

    pub(crate) fn get_dead_players<T: IntoIterator<Item = ClientId>>(
        &self,
        all_clients: T,
    ) -> impl Iterator<Item = ClientId> {
        all_clients
            .into_iter()
            .filter(|client_id| self.get_player_status(*client_id) != PlayerStatus::Alive)
    }

    fn end_cycle(&mut self) -> &mut Self {
        tracing::info!("ending cycle with votes: {:?}", self.votes);

        match self.cycle {
            Cycle::Day => {
                let num_votes_for_player =
                    self.votes
                        .iter()
                        .fold(HashMap::new(), |mut acc, (_, &target)| {
                            if let Some(target) = target {
                                *acc.entry(target).or_insert(0) += 1;
                            }

                            acc
                        });

                let num_players_alive = self.get_alive_players().count();

                if let Some((voted_player, _)) = num_votes_for_player
                    .into_iter()
                    .find(|(_, count)| count * 2 > num_players_alive)
                {
                    tracing::info!("{:?} was killed during the day", voted_player);
                    // TODO(emersonford): add event for vote result / death
                    *self
                        .player_status
                        .get_mut(&voted_player)
                        .expect("valid player") = PlayerStatus::Dead;
                }

                // TODO(emersonford): add event for failed vote result
            }
            Cycle::Night => {
                let protected_players = self.role_to_players.get(&SpecialRole::Doctor).map_or_else(
                    HashSet::new,
                    |players| {
                        players
                            .iter()
                            .filter_map(|client_id| self.votes.get(client_id))
                            .flatten()
                            .copied()
                            .collect::<HashSet<_>>()
                    },
                );

                let num_mafia_votes_for_player = self
                    .votes
                    .iter()
                    .filter(|(voter, _)| self.get_player_allegiance(**voter) == Allegiance::Mafia)
                    .fold(HashMap::new(), |mut acc, (_, &target)| {
                        if let Some(target) = target {
                            *acc.entry(target).or_insert(0) += 1;
                        }
                        acc
                    });

                let num_mafia_alive = self
                    .get_alive_players()
                    .filter(|client_id| self.get_player_allegiance(*client_id) == Allegiance::Mafia)
                    .count();

                if let Some((mafia_voted_player, _)) = num_mafia_votes_for_player
                    .into_iter()
                    .find(|(_, count)| count * 2 > num_mafia_alive)
                {
                    // TODO(emersonford): add event for vote result / death
                    if !protected_players.contains(&mafia_voted_player) {
                        tracing::info!(
                            "{:?} was killed by the mafia in the night",
                            mafia_voted_player
                        );
                        *self
                            .player_status
                            .get_mut(&mafia_voted_player)
                            .expect("valid player") = PlayerStatus::Dead;
                    } else {
                        tracing::info!(
                            "{:?} was protected from a mafia kill in the night",
                            mafia_voted_player
                        );
                    }
                }
                // TODO(emersonford): add event for failed vote result

                for investigator in self
                    .role_to_players
                    .get(&SpecialRole::Detective)
                    .into_iter()
                    .flatten()
                {
                    if let Some(target) = self.votes.get(investigator).copied().flatten() {
                        let allegiance = self.get_player_allegiance(target);

                        tracing::info!(
                            "{:?} was investigated by {:?} and discovered to be {:?}",
                            target,
                            investigator,
                            allegiance
                        );

                        // TODO(emersonford): add event for investigation
                    }
                }
            }
            _ => {
                return self;
            }
        }

        let num_mafia_alive = self
            .get_alive_players()
            .filter(|client_id| self.get_player_allegiance(*client_id) == Allegiance::Mafia)
            .count();

        if num_mafia_alive == 0 {
            tracing::info!("all mafia eliminated, villagers win");
            // TODO(emersonford): add event for villager win
            self.cycle = Cycle::Won(Allegiance::Villagers);
            return self;
        }

        let num_players_alive = self.get_alive_players().count();

        if num_mafia_alive * 2 >= num_players_alive {
            tracing::info!("#mafia >= #non mafia; mafia win");
            // TODO(emersonford): add event for mafia win
            self.cycle = Cycle::Won(Allegiance::Mafia);
            return self;
        }

        if self.day_num >= 100 {
            tracing::error!("game exceeded 100 rounds, defaulting win to mafia");
            // TODO(emersonford): add event for mafia win
            self.cycle = Cycle::Won(Allegiance::Mafia);
            return self;
        }

        self.votes = HashMap::new();
        self.cycle = self.cycle.next();
        self.day_num = if matches!(self.cycle, Cycle::Day) {
            self.day_num + 1
        } else {
            self.day_num
        };

        self
    }

    #[tracing::instrument(
        skip_all,
        fields(
            cycle = format!("{:?} {}", self.cycle, self.day_num),
            voter = voter.0,
            target = field::debug(target.map(|v| v.0)),
        )
    )]
    pub(crate) fn cast_vote(
        &mut self,
        voter: ClientId,
        target: Option<ClientId>,
    ) -> Result<&mut Self, MafiaGameError> {
        if self.get_player_status(voter) != PlayerStatus::Alive {
            return Err(MafiaGameError::InvalidVote(format!(
                "voter {:?} is not alive",
                voter
            )));
        }

        if target.is_some_and(|t| self.get_player_status(t) != PlayerStatus::Alive) {
            return Err(MafiaGameError::InvalidVote(format!(
                "target for vote {:?} is not alive",
                target
            )));
        }

        if SystemTime::now()
            .duration_since(self.cycle_start)
            .unwrap_or(Duration::from_secs(0))
            < self.config.vote_grace_period
        {
            return Err(MafiaGameError::InvalidVote(format!(
                "must wait {:?} after cycle start to cast vote",
                self.config.vote_grace_period
            )));
        }

        match self.cycle {
            Cycle::Day => {
                // TODO(emersonford): add event for vote cast
                self.votes.insert(voter, target);
            }
            Cycle::Night => {
                if !self.player_to_role.get(&voter).is_some_and(|v| {
                    matches!(
                        v,
                        SpecialRole::Mafia | SpecialRole::Doctor | SpecialRole::Detective
                    )
                }) {
                    return Err(MafiaGameError::InvalidVote(format!(
                        "{:?} does not have a role eligible to vote in {:?}",
                        voter, self.cycle
                    )));
                }

                // TODO(emersonford): add event for vote cast
                self.votes.insert(voter, target);
            }
            Cycle::Won(_) => {
                return Err(MafiaGameError::InvalidVote("game is complete".to_string()));
            }
        }

        Ok(self.poll_end_cycle())
    }

    /// Checks if we've met the conditions to end the cycle, and if so, ends the cycle.
    #[tracing::instrument(
        skip(self),
        fields(cycle = format!("{:?} {}", self.cycle, self.day_num)),
    )]
    pub(crate) fn poll_end_cycle(&mut self) -> &mut Self {
        if matches!(self.cycle, Cycle::Won(_)) {
            return self;
        }

        if SystemTime::now()
            .duration_since(self.cycle_start)
            .unwrap_or(Duration::from_secs(0))
            > if self.cycle == Cycle::Day {
                self.config.time_for_day
            } else {
                self.config.time_for_night
            }
        {
            tracing::info!("reached cycle end time, ending cycle");
            return self.end_cycle();
        }

        if self.cycle == Cycle::Day && self.config.end_day_after_all_votes {
            let num_players_alive = self.get_alive_players().count();

            if self.votes.len() == num_players_alive {
                tracing::info!("all votes cast, ending cycle");
                return self.end_cycle();
            }
        }

        if self.cycle == Cycle::Night && self.config.end_night_after_all_votes {
            let num_special_roles_alive = self
                .get_alive_players()
                .filter(|client_id| self.player_to_role.contains_key(client_id))
                .count();

            if self.votes.len() == num_special_roles_alive {
                tracing::info!("all votes cast, ending cycle");
                return self.end_cycle();
            }
        }

        self
    }

    pub(crate) fn get_cycle(&self) -> Cycle {
        self.cycle
    }
}
