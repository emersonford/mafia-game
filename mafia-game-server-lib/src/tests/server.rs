use std::collections::HashMap;
use std::time::Duration;

use insta::assert_json_snapshot;
use rand::rngs::mock::StepRng;

use crate::MafiaGameServer;
use crate::MafiaGameServerConfig;
use crate::game::GameConfig;
use mafia_game_lib::Allegiance;
use mafia_game_lib::Cycle;
use mafia_game_lib::SpecialRole;

#[test_log::test]
fn test_server_messages() {
    let server = MafiaGameServer::new(MafiaGameServerConfig {
        max_client_inactive_time: Duration::from_secs(300),
        randomize_death_message: false,
    });

    let (client0_id, client0_token) = server.connect_client("garnet").unwrap();
    let (client1_id, client1_token) = server.connect_client("amethyst").unwrap();
    let (client2_id, client2_token) = server.connect_client("pearl").unwrap();
    let (client3_id, client3_token) = server.connect_client("steven").unwrap();
    let (_client4_id, client4_token) = server.connect_client("connie").unwrap();
    let (_client5_id, client5_token) = server.connect_client("pink").unwrap();
    let (client6_id, client6_token) = server.connect_client("blue").unwrap();

    server.broadcast_message(Box::from("game is starting!"));

    server
        .start_game(
            GameConfig {
                start_cycle: Cycle::Day,
                time_for_day: Duration::from_secs(10),
                end_day_after_all_votes: true,
                time_for_night: Duration::from_secs(10),
                end_night_after_all_votes: true,
                num_special_roles: HashMap::from_iter([
                    (SpecialRole::Mafia, 2),
                    (SpecialRole::Detective, 1),
                    (SpecialRole::Doctor, 1),
                ]),
                vote_grace_period: Duration::from_secs(0),
            },
            StepRng::new(1, 1),
        )
        .unwrap();

    // Joined after the game started.
    let (_client7_id, client7_token) = server.connect_client("yellow").unwrap();

    assert_eq!(
        *server
            .0
            .read()
            .unwrap()
            .active_game
            .as_ref()
            .unwrap()
            .get_player_roles(),
        HashMap::from_iter([
            (client6_id, SpecialRole::Mafia),
            (client0_id, SpecialRole::Mafia),
            (client1_id, SpecialRole::Doctor),
            (client2_id, SpecialRole::Detective)
        ]),
    );

    // -- DAY 1 --
    server
        .send_message(client0_token, "hey everyone!".into())
        .unwrap();
    server
        .send_message(client1_token, "hey hey".into())
        .unwrap();
    server.send_message(client3_token, "hi!".into()).unwrap();
    server
        .send_message(client7_token, "oh I joined late :(".into())
        .unwrap();

    server.cast_vote(client0_token, None).unwrap();
    server.cast_vote(client1_token, None).unwrap();
    server.cast_vote(client2_token, None).unwrap();
    server.cast_vote(client3_token, None).unwrap();
    server.cast_vote(client4_token, None).unwrap();
    server.cast_vote(client5_token, None).unwrap();
    server.cast_vote(client6_token, None).unwrap();
    server.cast_vote(client7_token, None).unwrap_err();

    // -- NIGHT 1 --
    server
        .send_message(client6_token, "let's kill steven".into())
        .unwrap();
    server.send_message(client0_token, "okay".into()).unwrap();
    server
        .send_message(client1_token, "wonder what's gonna happen tonight".into())
        .unwrap();
    server
        .send_message(client7_token, "looks like fun!".into())
        .unwrap();

    server.cast_vote(client6_token, Some(client3_id)).unwrap();
    server.cast_vote(client0_token, Some(client3_id)).unwrap();

    // Joined after the game started.
    let (_client8_id, client8_token) = server.connect_client("white").unwrap();

    server.cast_vote(client1_token, None).unwrap();
    server.cast_vote(client2_token, Some(client6_id)).unwrap();

    // -- DAY 2 --
    server.send_message(client3_token, "wtf".into()).unwrap();
    server
        .send_message(client7_token, "welcome to the club".into())
        .unwrap();
    server
        .send_message(client2_token, "detective here, blue is mafia".into())
        .unwrap();
    server
        .send_message(client0_token, "how can we trust you?".into())
        .unwrap();
    server.send_message(client2_token, "idk".into()).unwrap();
    server
        .send_message(client4_token, "seems legit".into())
        .unwrap();
    server
        .send_message(client6_token, "hold on!!".into())
        .unwrap();

    server.cast_vote(client0_token, Some(client2_id)).unwrap();
    server.cast_vote(client1_token, Some(client6_id)).unwrap();
    server.cast_vote(client2_token, Some(client6_id)).unwrap();
    server.cast_vote(client3_token, None).unwrap_err();
    server.cast_vote(client4_token, Some(client6_id)).unwrap();
    server.cast_vote(client5_token, Some(client6_id)).unwrap();
    server.cast_vote(client6_token, Some(client2_id)).unwrap();
    server.cast_vote(client7_token, None).unwrap_err();

    // -- NIGHT 2 --
    server.send_message(client6_token, "damn".into()).unwrap();
    server
        .send_message(client0_token, "sorry blue, i tried".into())
        .unwrap();
    server
        .send_message(client3_token, "you killed me?".into())
        .unwrap();
    server
        .send_message(client2_token, "garnet seems sus".into())
        .unwrap();
    server
        .send_message(client6_token, "nothing personal".into())
        .unwrap();
    server
        .send_message(client1_token, "imma protect pearl".into())
        .unwrap();

    server.cast_vote(client0_token, Some(client2_id)).unwrap();
    server.cast_vote(client1_token, Some(client2_id)).unwrap();
    server.cast_vote(client2_token, Some(client0_id)).unwrap();

    // -- DAY 3 --
    server
        .send_message(client7_token, "tragic lol".into())
        .unwrap();
    server
        .send_message(client5_token, "wow good job doctor!".into())
        .unwrap();
    server
        .send_message(client2_token, "garnet is the other mafia".into())
        .unwrap();
    server
        .send_message(client6_token, "oh we lost rip".into())
        .unwrap();
    server.send_message(client0_token, "no!".into()).unwrap();

    server.cast_vote(client0_token, Some(client2_id)).unwrap();
    server.cast_vote(client1_token, Some(client0_id)).unwrap();
    server.cast_vote(client2_token, Some(client0_id)).unwrap();
    server.cast_vote(client3_token, None).unwrap_err();
    server.cast_vote(client4_token, Some(client0_id)).unwrap();
    server.cast_vote(client5_token, Some(client0_id)).unwrap();
    server.cast_vote(client6_token, None).unwrap_err();
    server.cast_vote(client7_token, None).unwrap_err();

    // -- VILLAGERS WIN --
    server.broadcast_message(Box::from("villagers won"));

    server.send_message(client0_token, "shit".into()).unwrap();
    server.send_message(client6_token, "gg".into()).unwrap();
    server
        .send_message(client2_token, "yeah gg all".into())
        .unwrap();
    server
        .send_message(client7_token, "hope to join the next one!".into())
        .unwrap();

    assert_eq!(
        server
            .0
            .read()
            .unwrap()
            .active_game
            .as_ref()
            .unwrap()
            .get_winner(),
        Some(Allegiance::Villagers)
    );

    insta::with_settings!({sort_maps => true}, {
        assert_json_snapshot!(server.take_events(client0_token).unwrap());
        assert_json_snapshot!(server.take_events(client1_token).unwrap());
        assert_json_snapshot!(server.take_events(client2_token).unwrap());
        assert_json_snapshot!(server.take_events(client3_token).unwrap());
        assert_json_snapshot!(server.take_events(client4_token).unwrap());
        assert_json_snapshot!(server.take_events(client5_token).unwrap());
        assert_json_snapshot!(server.take_events(client6_token).unwrap());
        assert_json_snapshot!(server.take_events(client7_token).unwrap());
        assert_json_snapshot!(server.take_events(client8_token).unwrap());
    });
}
