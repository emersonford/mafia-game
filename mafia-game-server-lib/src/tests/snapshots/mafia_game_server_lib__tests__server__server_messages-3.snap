---
source: mafia-game-server-lib/src/tests/server.rs
expression: server.take_events(client2_token).unwrap()
---
[
  {
    "SetServerInfo": {
      "connected_clients": [
        {
          "name": "garnet",
          "id": 0
        },
        {
          "name": "amethyst",
          "id": 1
        },
        {
          "name": "pearl",
          "id": 2
        }
      ],
      "active_game": null
    }
  },
  {
    "ClientConnected": {
      "name": "steven",
      "id": 3
    }
  },
  {
    "ClientConnected": {
      "name": "connie",
      "id": 4
    }
  },
  {
    "ClientConnected": {
      "name": "pink",
      "id": 5
    }
  },
  {
    "ClientConnected": {
      "name": "blue",
      "id": 6
    }
  },
  {
    "MessageReceived": {
      "channel": "Public",
      "contents": "game is starting!",
      "from": "System"
    }
  },
  {
    "SetGame": {
      "cycle_start_time_unix_ts_secs": 0,
      "cycle_duration_secs": 10,
      "current_cycle": "Day",
      "day_num": 1,
      "player_to_role": {
        "2": "Detective"
      },
      "player_status": {
        "0": "Alive",
        "1": "Alive",
        "2": "Alive",
        "3": "Alive",
        "4": "Alive",
        "5": "Alive",
        "6": "Alive"
      },
      "votes": {},
      "winner": null
    }
  },
  {
    "ClientConnected": {
      "name": "yellow",
      "id": 7
    }
  },
  {
    "MessageReceived": {
      "channel": "Public",
      "contents": "hey everyone!",
      "from": {
        "Client": 0
      }
    }
  },
  {
    "MessageReceived": {
      "channel": "Public",
      "contents": "hey hey",
      "from": {
        "Client": 1
      }
    }
  },
  {
    "MessageReceived": {
      "channel": "Public",
      "contents": "hi!",
      "from": {
        "Client": 3
      }
    }
  },
  {
    "VoteIssued": {
      "voter": 0,
      "target": null,
      "channel": "Public"
    }
  },
  {
    "VoteIssued": {
      "voter": 1,
      "target": null,
      "channel": "Public"
    }
  },
  {
    "VoteIssued": {
      "voter": 2,
      "target": null,
      "channel": "Public"
    }
  },
  {
    "VoteIssued": {
      "voter": 3,
      "target": null,
      "channel": "Public"
    }
  },
  {
    "VoteIssued": {
      "voter": 4,
      "target": null,
      "channel": "Public"
    }
  },
  {
    "VoteIssued": {
      "voter": 5,
      "target": null,
      "channel": "Public"
    }
  },
  {
    "VoteIssued": {
      "voter": 6,
      "target": null,
      "channel": "Public"
    }
  },
  {
    "FailedVote": {
      "cycle": "Day",
      "channel": "Public"
    }
  },
  {
    "SetCycle": {
      "cycle": "Night",
      "day_num": 1
    }
  },
  {
    "ClientConnected": {
      "name": "white",
      "id": 8
    }
  },
  {
    "VoteIssued": {
      "voter": 2,
      "target": 6,
      "channel": "Spectator"
    }
  },
  {
    "PlayerKilled": {
      "player": 3,
      "cycle": "Night",
      "death_message": "was found stabbed to death"
    }
  },
  {
    "PlayerInvestigated": {
      "actor": 2,
      "target": 6,
      "allegiance": "Mafia"
    }
  },
  {
    "SetCycle": {
      "cycle": "Day",
      "day_num": 2
    }
  },
  {
    "MessageReceived": {
      "channel": "Public",
      "contents": "detective here, blue is mafia",
      "from": {
        "Client": 2
      }
    }
  },
  {
    "MessageReceived": {
      "channel": "Public",
      "contents": "how can we trust you?",
      "from": {
        "Client": 0
      }
    }
  },
  {
    "MessageReceived": {
      "channel": "Public",
      "contents": "idk",
      "from": {
        "Client": 2
      }
    }
  },
  {
    "MessageReceived": {
      "channel": "Public",
      "contents": "seems legit",
      "from": {
        "Client": 4
      }
    }
  },
  {
    "MessageReceived": {
      "channel": "Public",
      "contents": "hold on!!",
      "from": {
        "Client": 6
      }
    }
  },
  {
    "VoteIssued": {
      "voter": 0,
      "target": 2,
      "channel": "Public"
    }
  },
  {
    "VoteIssued": {
      "voter": 1,
      "target": 6,
      "channel": "Public"
    }
  },
  {
    "VoteIssued": {
      "voter": 2,
      "target": 6,
      "channel": "Public"
    }
  },
  {
    "VoteIssued": {
      "voter": 4,
      "target": 6,
      "channel": "Public"
    }
  },
  {
    "VoteIssued": {
      "voter": 5,
      "target": 6,
      "channel": "Public"
    }
  },
  {
    "VoteIssued": {
      "voter": 6,
      "target": 2,
      "channel": "Public"
    }
  },
  {
    "PlayerKilled": {
      "player": 6,
      "cycle": "Day",
      "death_message": "was hung for their unforgivable sins"
    }
  },
  {
    "SetCycle": {
      "cycle": "Night",
      "day_num": 2
    }
  },
  {
    "MessageReceived": {
      "channel": "Spectator",
      "contents": "garnet seems sus",
      "from": {
        "Client": 2
      }
    }
  },
  {
    "VoteIssued": {
      "voter": 2,
      "target": 0,
      "channel": "Spectator"
    }
  },
  {
    "PlayerInvestigated": {
      "actor": 2,
      "target": 0,
      "allegiance": "Mafia"
    }
  },
  {
    "SetCycle": {
      "cycle": "Day",
      "day_num": 3
    }
  },
  {
    "MessageReceived": {
      "channel": "Public",
      "contents": "wow good job doctor!",
      "from": {
        "Client": 5
      }
    }
  },
  {
    "MessageReceived": {
      "channel": "Public",
      "contents": "garnet is the other mafia",
      "from": {
        "Client": 2
      }
    }
  },
  {
    "MessageReceived": {
      "channel": "Public",
      "contents": "no!",
      "from": {
        "Client": 0
      }
    }
  },
  {
    "VoteIssued": {
      "voter": 0,
      "target": 2,
      "channel": "Public"
    }
  },
  {
    "VoteIssued": {
      "voter": 1,
      "target": 0,
      "channel": "Public"
    }
  },
  {
    "VoteIssued": {
      "voter": 2,
      "target": 0,
      "channel": "Public"
    }
  },
  {
    "VoteIssued": {
      "voter": 4,
      "target": 0,
      "channel": "Public"
    }
  },
  {
    "VoteIssued": {
      "voter": 5,
      "target": 0,
      "channel": "Public"
    }
  },
  {
    "PlayerKilled": {
      "player": 0,
      "cycle": "Day",
      "death_message": "was hung for their unforgivable sins"
    }
  },
  {
    "GameWon": {
      "player_to_role": {
        "0": "Mafia",
        "1": "Doctor",
        "2": "Detective",
        "6": "Mafia"
      },
      "side": "Villagers"
    }
  },
  {
    "MessageReceived": {
      "channel": "Public",
      "contents": "villagers won",
      "from": "System"
    }
  },
  {
    "MessageReceived": {
      "channel": "Public",
      "contents": "shit",
      "from": {
        "Client": 0
      }
    }
  },
  {
    "MessageReceived": {
      "channel": "Public",
      "contents": "gg",
      "from": {
        "Client": 6
      }
    }
  },
  {
    "MessageReceived": {
      "channel": "Public",
      "contents": "yeah gg all",
      "from": {
        "Client": 2
      }
    }
  },
  {
    "MessageReceived": {
      "channel": "Public",
      "contents": "hope to join the next one!",
      "from": {
        "Client": 7
      }
    }
  }
]
